### DISCUSSION TITLE
Help with parsing whitespace-sensitive expressions using external scanners

### DISCUSSION THREAD

**Discussion by ahelwer** [tree-sitter/tree-sitter#1160] [Q&A] [answered]
This is my first time trying to use an external scanner to parse something. The language [TLA+](https://en.wikipedia.org/wiki/TLA%2B) includes an easy-to-read method of writing conjunctions & disjunctions using bulleted vertical alignment; for example:
```
e ==
    /\ A
    /\ B
    /\  \/ C
        \/ D
    /\ E
```
where `/\` is conjunction and `\/` is disjunction; the above expression should parse to `A /\ B /\ (C \/ D) /\ E`, or something like:
```
(op_def
  (conj_list
    (identifier)
    (identifier)
    (disj_list
      (identifier)
      (identifier)
    )
    (identifier)
  )
)
```
At this point from looking at the Python tree-sitter grammar (and taking a stab at it myself) I have a pretty good grasp of how to write an external scanner, what its capabilities are, etc. The issue is the interface between the grammar itself and the scanner; what tokens should I use and how should I define my `conj_list` rule to match the semantics I want? I'm trying something like:
```
conj_list: $ => seq(
  $._indent, $._conj_bullet, $._expr,
  repeat(seq($._newline, $._indent, $._conj_bullet, $._expr))
)
```
where everything except `_expr` is an external token. The `_indent` token is emitted if the scanner sees 0 or more whitespace followed by a `/\` token, and the `_conj_bullet` token is emitted upon seeing a `/\` token (can possible move this into the grammar itself). Basic desired semantics (focusing only on conjugation for simplicity, although disjunction follows the same rules):

- if the next `_conj_bullet` is in the same column position as the first bullet in the list, that bullet should be added to the `conj_list`
- if the next `_conj_bullet` is in a column position greater than the position of the first bullet in the list, that should start a new nested `conj_list`
- if any non-whitespace token appears in a column before the position of the first bullet in the list, that ends the list

Any help would be greatly appreciated!

**Comment by razzeee** [comment]
I'm wondering if your first line also needs a `_newline`? That way you could could down on logic and possibly join `_newline` and `_indent`.

Also if you can, move as much as possible to the grammar itself, usually if something is a pain/hard to maintain, it's due to in being in the external scanner.

**Comment by ahelwer** [comment, marked as answer]
I haven't fully accounted for every single case yet but I'm having great success with `INDENT`, `NEWLINE`, and `DEDENT` external tokens; these mark the list beginning, the separation between each list element, and the list end respectively  (their names really aren't too important). This seems like a very general way to parse lists with arbitrary lookahead & backward-looking state.
```
conj_list: $ => seq(
  $._indent, $.conj_item,
  repeat(seq($._newline, $.conj_item)),
  $._dedent
),

conj_item: $ => seq($.land, $._expr),

land: $ => choice('/\\', '∧')
```
To add to the intrigue, `land` also works as an infix operator in this language! Many cases to handle. Having fun though.
